pkgs/applications/science/logic/coq/default.nix

```diff
@@ -5,57 +5,76 @@
-  ideFlags = stdenv.lib.optionalString (buildIde && !versionAtLeast "8.10")
+  release = {
+   "8.5pl1".sha256     = "1976ki5xjg2r907xj9p7gs0kpdinywbwcqlgxqw75dgp0hkgi00n";
+   "8.5pl2".sha256     = "109rrcrx7mz0fj7725kjjghfg5ydwb24hjsa5hspa27b4caah7rh";
+   "8.5pl3".sha256     = "15c3rdk59nifzihsp97z4vjxis5xmsnrvpb86qiazj143z2fmdgw";
+   "8.6.0".sha256      = "148mb48zpdax56c0blfi7v67lx014lnmrvxxasi28hsibyz2lvg4";
+   "8.6.0".rev         = "V8.6";
+   "8.6.1".sha256      = "0llrxcxwy5j87vbbjnisw42rfw1n1pm5602ssx64xaxx3k176g6l";
+   "8.7.0".sha256      = "1h18b7xpnx3ix9vsi5fx4zdcbxy7bhra7gd5c5yzxmk53cgf1p9m";
+   "8.7.1".sha256      = "0gjn59jkbxwrihk8fx9d823wjyjh5m9gvj9l31nv6z6bcqhgdqi8";
+   "8.7.2".sha256      = "0a0657xby8wdq4aqb2xsxp3n7pmc2w4yxjmrb2l4kccs1aqvaj4w";
+   "8.8.0".sha256      = "13a4fka22hdxsjk11mgjb9ffzplfxyxp1sg5v1c8nk1grxlscgw8";
+   "8.8.1".sha256      = "1hlf58gwazywbmfa48219amid38vqdl94yz21i11b4map6jfwhbk";
+   "8.8.2".sha256      = "1lip3xja924dm6qblisk1bk0x8ai24s5xxqxphbdxj6djglj68fd";
+   "8.9.0".sha256      = "1dkgdjc4n1m15m1p724hhi5cyxpqbjw6rxc5na6fl3v4qjjfnizh";
+   "8.9.1".sha256      = "1xrq6mkhpq994bncmnijf8jwmwn961kkpl4mwwlv7j3dgnysrcv2";
+   "8.10.0".sha256     = "138jw94wp4mg5dgjc2asn8ng09ayz1mxdznq342n0m469j803gzg";
+   "8.10.1".sha256     = "072v2zkjzf7gj48137wpr3c9j0hg9pdhlr5l8jrgrwynld8fp7i4";
+   "8.10.2".sha256     = "0znxmpy71bfw0p6x47i82jf5k7v41zbz9bdpn901ysn3ir8l3wrz";
+   "8.11.0".sha256     = "1rfdic6mp7acx2zfwz7ziqk12g95bl9nyj68z4n20a5bcjv2pxpn";
+   "8.11.1".sha256     = "0qriy9dy36dajsv5qmli8gd6v55mah02ya334nw49ky19v7518m0";
+   "8.11.2".sha256     = "0f77ccyxdgbf1nrj5fa8qvrk1cyfy06fv8gj9kzfvlcgn0cf48sa";
+   "8.12.0".sha256     = "18dc7k0piv6v064zgdadpw6mkkxk7j663hb3svgj5236fihjr0cz";
+   "8.12.1".sha256     = "1rkcyjjrzcqw9xk93hsq0vvji4f8r5iq0f739mghk60bghkpnb7q";
+   "8.12.2".sha256     = "18gscfm039pqhq4msq01nraig5dm9ab98bjca94zldf8jvdv0x2n";
+   "8.13+beta1".sha256 = "1v4a6dpj41flspa4ihcr7m5ahqz10kbn62fmrldmv7gzq6jsyfyq";
+  };
+  releaseRev = v: "V${v}";
+  fetched = import ../../../../build-support/coq/meta-fetch/default.nix
+    {inherit (stdenv) lib; inherit fetchzip; }
+    { inherit release releaseRev;
+      location = { owner = "coq"; repo = "coq"; }; } args.version;
+  version = fetched.version;
+  coq-version = if version != "dev" then versions.majorMinor version else "dev";
+  versionAtLeast = v: (version == "dev") || (stdenv.lib.versionAtLeast coq-version v);
+  ideFlags = optionalString (buildIde && !versionAtLeast "8.10")
     "-lablgtkdir ${ocamlPackages.lablgtk}/lib/ocaml/*/site-lib/lablgtk2 -coqide opt";
   csdpPatch = if csdp != null then ''
     substituteInPlace plugins/micromega/sos.ml --replace "; csdp" "; ${csdp}/bin/csdp"
     substituteInPlace plugins/micromega/coq_micromega.ml --replace "System.is_in_system_path \"csdp\"" "true"
   '' else "";
+  ocamlPackages = if isNull (args.ocamlPackages or null) then args.ocamlPackages else
```

Comment:
I don’t understand this line.

The `mkCoq` function below always set the `ocamlPackages` argument, so it is not null, therefore it is ignored and the default `ocamlPackages` set is computed depending on `coq-version`.

Suggestion: move the logic for computing the right version of OCaml out of the coq derivation (it thus becomes straightforward to overwrite).

Reply:
The `mkCoq` function is a all to `callPackage ../applications/science/logic/coq` which leaves out `coqPackages` indeed, but `callPackage` provides the ability to override the arguments and hence one can write `coq.override {ocamlPackages = whatever;}` to change the version of ocaml used by Coq.

Reply:
oh, I did two mistakes that cancel each other, sorry! And well spot!
I'll push a fix asap.

Reply:
I fixed my mistake and documented the new instructions... which are using an attribute `customOcamlPackages`.

Reply:
https://github.com/NixOS/nixpkgs/blob/14ee56c849763beafce003258fa7ce2ba9512599/doc/languages-frameworks/coq.section.md

Reply:
I feel like you are inventing yet an other overriding mechanism. I’m glad you did not duplicate all parameters (`make` & `customMake`, etc.). But I won’t argue against it.

Thanks for the fix nonetheless.

Reply:
There are two reasons combined that made me duplicate this predicate that would not possibly make me duplicate any other. 1. it's a scoped argument, which is passed implicitly by callPackage, so I cannot prevent it from being always set whereas I wanted it to be optional. 2. it's the only argument that is both a derivation and needs to change depending on the Coq version.

I don't see how this is yet another overriding mechanism, I'm merly reusing `.override` provided by `callPackage`.

Do you see another solution that would make it possible to have the following two derivation build correctly:
- `coq.override {version = "8.10"};`
- `coq.override {version = "d370a9d1328a4e1cdb9d02ee032f605a9d94ec7a"; customOcamlPackages = ocaml-ng.ocamlPackages_4_9}`;

Now I tried thoses two commands, I realize that the second one does not work anymore since I started using `fetchzip` everywhere. See my next comment.

Reply:
What I meant about this overriding scheme being ad-hoc, is that there is an argument whose only purpose is overriding. Notice that now there are two ways to override the ocamlPackages set: either by setting `customOCamlPackages` or by overriding, say, `ocamlPackages_4_10`.